### ISSUE TITLE
请问为什么不去切换motion的数据🤪这样不会有数据没训到吗?

### ISSUE THREAD

**Issue by toonasinensis** [Axellwppr/motion_tracking#16]
# [Possible Bug / Design Question] Motion IDs are sampled only once and remain fixed across resets

# 【可能的缺陷 / 设计确认】Motion ID 只在初始化时采样一次，reset 后始终绑定原 motion

### 问题摘要

`ProgressiveMultiMotionDataset` 在构造时为每个并行环境采样一次 motion ID，但正常 episode reset、motion timeout 和失败后的 reset 都不会重新采样 motion。Reset 只会改变该 motion 内部的局部帧 `t`。

这可能是有意设计，也可能会导致大数据集训练覆盖率明显不足。希望确认预期行为。

### 当前行为

构造数据集时会为所有 env 执行一次采样：

```python
self._all_env_ids = torch.arange(self.env_size, device=self.device, dtype=torch.long)
self._resample_motion_ids(self._all_env_ids)
self._invalidate_cache(self._all_env_ids)
```

但 `reset()` 中的重新采样和 cache 失效代码被注释：

```python
def reset(self, env_ids: torch.Tensor) -> torch.Tensor:
    env_ids = env_ids.long()
    # do not resample motion ids or invalidate cache here
    # self._resample_motion_ids(env_ids)
    # self._invalidate_cache(env_ids)
    return self._len_A[env_ids]
```

调用方虽然有 `# resample motion` 注释，但实际调用的上述 `dataset.reset()` 不会重采样：

```python
def sample_init(self, env_ids: torch.Tensor):
    t = self.t[env_ids]
    # resample motion
    lengths = self.dataset.reset(env_ids)
    ...
    t_rand = (torch.rand(n, device=self.device) * sample_interval).floor()
    self.t[env_ids] = torch.where(use_reinit, reinit_t, t_rand)
```

所以实际行为是：

```text
dataset构造：每个env抽一条motion
episode reset：motion ID不变，只重新选择local frame t
motion timeout：motion ID不变，只重新选择local frame t
recovery reset：motion ID不变，通常将local frame t向前回退
```

### 本地长时间运行验证

验证环境：

```text
GPU：NVIDIA GeForce RTX 4090
并行环境：4000
计划PPO iteration：4000
数据集：roban/seed_train
数据集motion数量：115481
验证脚本：roban_tools/verify_motion_lock_train.py
```

验证器没有改变采样结果或 reset 返回值，只做以下旁路统计：

1. 统计 `_resample_motion_ids()` 的总调用次数。
2. 在每次 `dataset.reset(env_ids)` 前后比较对应 env 的 motion ID。
3. 每60秒将当前全部 motion ID 与构造完成后的初始 ID 比较。

运行超过1小时后的快照：

```json
{
  "elapsed_s": 3783.506,
  "num_envs": 4000,
  "num_motions": 115481,
  "resample_calls_total": 1,
  "reset_calls": 38170,
  "reset_env_events": 1013264,
  "motion_id_changes_during_reset": 0,
  "motion_ids_different_from_initial": 0,
  "all_motion_ids_stable": true
}
```

此时训练已完成约 `1202/4000` 个 PPO iteration。超过 `1,013,264` 次 env-reset 事件后：

- `_resample_motion_ids()` 仍只调用过1次，即数据集构造阶段。
- 没有任何 reset 导致 motion ID 改变。
- 4000个 env 当前的 motion ID 全部与初始化时相同。

### 对数据覆盖率的影响

Roban转换后的训练集包含：

```text
motion数量 M = 115481
```

采样使用：

```python
torch.multinomial(motion_sample_probs, num_envs, replacement=True)
```

因此是有放回采样。对于 `motion_uniform`，一次初始化中期望覆盖的不同 motion 数为：

```text
E[unique] = M × (1 - (1 - 1/M)^N)
```

| 配置 | env槽位 N | 不重复时的覆盖上限 | 有放回采样的期望不同motion | 期望覆盖率 |
|---|---:|---:|---:|---:|
| 本地验证 | 4,000 | 3.46% | 约3,932 | 3.40% |
| 当前4卡训练 | 4 × 6,144 = 24,576 | 21.28% | 约22,137 | 19.17% |
| 原项目建议规模 | 8 × 8,192 = 65,536 | 56.75% | 约50,011 | 43.31% |

由于 reset 不重新采样，单个训练阶段内覆盖率不会随着训练时间或 reset 次数继续增加。

### 期望行为需要确认

请确认以下哪一种才是预期设计：

1. **固定绑定是有意设计。** 每个 env 在整个训练阶段始终绑定同一条 motion，通过大量并行 env 提供数据多样性。若是这样，建议在文档中明确说明，并考虑数据集远大于 env 数量时的覆盖策略。
2. **普通 reset 应重新采样。** 正常 episode reset 和 motion timeout 重新抽 motion；只有 recovery/reinitialization 保留当前 motion 并回退时间。
3. **采用其他覆盖机制。** 例如按 rank 分片、无放回轮换或 epoch sampler，而不改变 recovery 语义。

本 issue 只报告和验证当前行为，不包含采样语义修改。

### 相关文件

- `active_adaptation/utils/multimotion.py`
- `active_adaptation/envs/mdp/commands/motion_tracking/core.py`
- `roban_tools/verify_motion_lock_train.py`
- `artifacts/motion_lock_4000env_4000batches/report.jsonl`

---

### Summary

`ProgressiveMultiMotionDataset` samples one motion ID for each parallel environment during construction. Normal episode resets, motion timeouts, and failure-triggered resets do not resample the motion. A reset only changes the local frame index `t` within the same motion.

This may be intentional, but it can severely limit coverage when the motion dataset is much larger than the number of parallel environments. We would like to confirm the intended behavior.

### Current behavior

All environment-to-motion assignments are sampled once during dataset construction:

```python
self._all_env_ids = torch.arange(self.env_size, device=self.device, dtype=torch.long)
self._resample_motion_ids(self._all_env_ids)
self._invalidate_cache(self._all_env_ids)
```

The resampling and cache invalidation calls are commented out in `reset()`:

```python
def reset(self, env_ids: torch.Tensor) -> torch.Tensor:
    env_ids = env_ids.long()
    # do not resample motion ids or invalidate cache here
    # self._resample_motion_ids(env_ids)
    # self._invalidate_cache(env_ids)
    return self._len_A[env_ids]
```

The caller contains a `# resample motion` comment, but the invoked `dataset.reset()` does not actually resample:

```python
def sample_init(self, env_ids: torch.Tensor):
    t = self.t[env_ids]
    # resample motion
    lengths = self.dataset.reset(env_ids)
    ...
    t_rand = (torch.rand(n, device=self.device) * sample_interval).floor()
    self.t[env_ids] = torch.where(use_reinit, reinit_t, t_rand)
```

The effective behavior is therefore:

```text
Dataset construction: sample one motion for every env
Episode reset: keep the motion ID and select a new local frame t
Motion timeout: keep the motion ID and select a new local frame t
Recovery reset: keep the motion ID and usually rewind local frame t
```

### Long-running local verification

Verification setup:

```text
GPU: NVIDIA GeForce RTX 4090
Parallel environments: 4,000
Planned PPO iterations: 4,000
Dataset: roban/seed_train
Number of motions: 115,481
Audit wrapper: roban_tools/verify_motion_lock_train.py
```

The audit wrapper does not alter sampling results or reset return values. It only:

1. Counts all calls to `_resample_motion_ids()`.
2. Compares motion IDs before and after every `dataset.reset(env_ids)` call.
3. Compares all current motion IDs with their post-construction values every 60 seconds.

Snapshot after more than one hour:

```json
{
  "elapsed_s": 3783.506,
  "num_envs": 4000,
  "num_motions": 115481,
  "resample_calls_total": 1,
  "reset_calls": 38170,
  "reset_env_events": 1013264,
  "motion_id_changes_during_reset": 0,
  "motion_ids_different_from_initial": 0,
  "all_motion_ids_stable": true
}
```

The run had completed approximately `1202/4000` PPO iterations at this point. After more than `1,013,264` per-environment reset events:

- `_resample_motion_ids()` had still been called exactly once, during dataset construction.
- No reset changed any motion ID.
- All 4,000 current motion IDs were identical to their initial values.

### Dataset coverage impact

The converted Roban training dataset contains:

```text
Number of motions M = 115,481
```

Sampling is performed with replacement:

```python
torch.multinomial(motion_sample_probs, num_envs, replacement=True)
```

For `motion_uniform`, the expected number of unique motions in one initialization is:

```text
E[unique] = M × (1 - (1 - 1/M)^N)
```

| Configuration | Env slots N | Maximum coverage without duplicates | Expected unique motions with replacement | Expected coverage |
|---|---:|---:|---:|---:|
| Local verification | 4,000 | 3.46% | ~3,932 | 3.40% |
| Current four-GPU run | 4 × 6,144 = 24,576 | 21.28% | ~22,137 | 19.17% |
| Upstream recommended scale | 8 × 8,192 = 65,536 | 56.75% | ~50,011 | 43.31% |

Because resets do not resample, coverage within one training stage does not increase with training duration or reset count.

### Intended behavior to clarify

Please confirm which behavior is intended:

1. **Fixed assignment is intentional.** Each env remains assigned to one motion for the entire training stage, and diversity comes from a large number of parallel envs. If so, this should be documented, together with a coverage strategy for datasets much larger than the env count.
2. **Normal resets should resample.** Normal episode resets and motion timeouts sample a new motion, while recovery/reinitialization keeps the current motion and rewinds time.
3. **Use another coverage mechanism.** Examples include per-rank sharding, without-replacement rotation, or an epoch sampler that preserves recovery semantics.

This issue only reports and verifies the current behavior. It does not include a change to sampling semantics.

### Relevant files

- `active_adaptation/utils/multimotion.py`
- `active_adaptation/envs/mdp/commands/motion_tracking/core.py`
- `roban_tools/verify_motion_lock_train.py`
- `artifacts/motion_lock_4000env_4000batches/report.jsonl`

**Comment by Axellwppr** [comment]
Thanks for the detailed investigation. You are right that the current implementation does not resample motion IDs after initialization.

Our no-mirror Seed dataset contains about 61k motion segments, while our recommended setup uses 8 × 8,192 = 65,536 environments. They are in the same order of magnitude. However, this does not guarantee full coverage because sampling is weighted across datasets and performed with replacement. In practice, our datasets already provide more motion diversity than we usually need for the tracking task, so we did not expect this issue to have a significant impact.

We apologize that this detail was not handled or documented carefully. The class is named `ProgressiveMultiMotionDataset` because it originally supported an optional progressive motion rotation mechanism. This mechanism was later temporarily disabled due to cache-management considerations. We will document and address this issue in the next release.

If needed, you can manually uncomment the resampling code. However, you will need to handle the recovery/rewind case carefully, since the local time t is intentionally moved backward within the current motion.
